src/eduid_userdb/credentials/fido.py

```diff
@@ -34,356 +34,100 @@
 #
 from __future__ import annotations
 
-import copy
-from datetime import datetime
+from dataclasses import dataclass
 from hashlib import sha256
-from typing import Any, Dict, Optional, Type, Union
-
-from six import string_types
+from typing import Any, ClassVar, Dict, Optional
 
 from eduid_userdb.credentials import Credential
-from eduid_userdb.exceptions import UserDBValueError, UserHasUnknownData
 
 __author__ = 'ft'
 
 
-class FidoCredential(Credential):
+@dataclass
+class _FidoCredentialRequired:
     """
-    Token authentication credential
+    Required fields for FidoCredential, so that they go before the optional
+    arguments of Element in the implicit constructor.
     """
 
-    def __init__(self, data: Dict[str, Any], called_directly: bool = True):
-
-        super().__init__(data, called_directly=called_directly)
-        self.keyhandle = data.pop('keyhandle')
-        self.app_id = data.pop('app_id')
-        self.description = data.pop('description', '')
-
-    def check_unknown_data(self, data: Dict[str, Any]):
-        """
-        called when an instance of a subclass is created with `raise_on_unknown`
-        """
-        leftovers = data.keys()
-        if leftovers:
-            raise UserHasUnknownData(f'{self.__class__.__name__} {self.key} unknown data: {leftovers}')
-
-    @property
-    def keyhandle(self):
-        """
-        This is the server side reference to the U2F token used.
-
-        :return: U2F keyhandle.
-        :rtype: str
-        """
-        return self._data['keyhandle']
-
-    @keyhandle.setter
-    def keyhandle(self, value):
-        """
-        :param value: U2F keyhandle.
-        :type value: str
-        """
-        if not isinstance(value, string_types):
-            raise UserDBValueError("Invalid 'keyhandle': {!r}".format(value))
-        self._data['keyhandle'] = value
+    keyhandle: str
+    app_id: str
 
-    @property
-    def app_id(self):
-        """
-        The U2F app_id used when creating this credential.
 
-        :return: U2F app_id
-        :rtype: str
-        """
-        return self._data['app_id']
+@dataclass(eq=False, unsafe_hash=True)
```

Comment:
Is there a reason that eq is set to False?

Would it not be better if keyhandle was used for a `__hash__` method instead of unsafe_hash?

Reply:
The main thing here is that 2 objects that evaluate as equal have to have the same hash. So here we are inheriting the `__hash__` and `__eq__` from [Credential](https://github.com/SUNET/eduid-userdb/blob/eperez-elem-init-4/src/eduid_userdb/credentials/base.py#L72).

Reply:
The above means that must guarantee that the hash will not change if the object changes: the `key` must be "immutable" - which I think is guaranteed.

Reply:
In any case, I want to add that I do not mean to push for making dataclasses hashable in this way; I am open to use frozen dataclasses. However, if we agree on using frozen dataclasses, I would do that in a further step, as a separate concern, since it would change the current contract and thus more code would change in a single PR. If we do it like this, we can keep for now the code that assumes that these objects are mutable; otherwise, that code will need to change, adding a further "context switch" to this PR.